File: packages/agent-sdk/src/channel-agent.ts
import type {
  Agent as UpstreamAgent,
  AuthenticateRequest,
  AuthenticateResponse,
  CancelNotification,
  InitializeRequest,
  InitializeResponse,
  NewSessionRequest,
  NewSessionResponse,
  PromptCapabilities,
  PromptRequest,
  PromptResponse,
  Stream,
} from '@agentclientprotocol/sdk'

import {AgentSideConnection, ndJsonStream} from '@agentclientprotocol/sdk'
import {randomUUID} from 'node:crypto'
import {Readable, Writable} from 'node:stream'

import {PromptContext} from './prompt-context.js'

export type ChannelAgentConfig = {
  readonly name: string
  readonly promptCapabilities: PromptCapabilities
  readonly version: string
}

export type ChannelAgentRunOptions = {
  /**
   * Optional bidirectional ACP `Stream`. Defaults to stdio NDJSON
   * (`ndJsonStream(stdout, stdin)`). Tests typically construct a paired
   * in-memory pair and pass it here.
   */
  readonly stream?: Stream
}

export type PromptHandler = (
  request: PromptRequest,
  ctx: PromptContext,
) => Promise<PromptResponse>

export type CancelHandler = (notification: CancelNotification) => Promise<void> | void

type SessionState = {
  readonly abortController: AbortController
}

/**
 * Ergonomic wrapper around `@agentclientprotocol/sdk`'s `AgentSideConnection`.
 *
 * Surface (Slice 5.1, v0.1):
 *   - `onPrompt(handler)` — register the user's prompt handler.
 *   - `onCancel(handler)` — register an optional cancel handler.
 *   - `run({stream?})` — wire stdin/stdout (or the provided stream) and
 *     start the agent loop.
 *
 * Outside-in: every API exists because the 25-LOC echo example needs it.
 * `loadSession`, `setSessionMode`, `authenticate` are intentionally absent
 * from v0.1 — add them only when an example surfaces the need.
 */
export class ChannelAgent {
  private cancelHandler: CancelHandler | undefined
  private readonly config: ChannelAgentConfig
  private connection: AgentSideConnection | undefined
  private promptHandler: PromptHandler | undefined
  private readonly sessions = new Map<string, SessionState>()

  public constructor(config: ChannelAgentConfig) {
    this.config = config
  }

  onCancel(handler: CancelHandler): void {
    this.cancelHandler = handler
  }

  onPrompt(handler: PromptHandler): void {
    this.promptHandler = handler
  }

  run(options: ChannelAgentRunOptions = {}): void {
    const stream = options.stream ?? this.defaultStdioStream()
    this.connection = new AgentSideConnection((conn) => this.makeAgent(conn), stream)
  }

  private defaultStdioStream(): Stream {
    // The upstream lib expects Web Streams. Node's process.stdout / stdin
    // are Node streams; we adapt them via the Node-built-in helpers.
    const stdoutWeb = Writable.toWeb(process.stdout) as WritableStream<Uint8Array>
    const stdinWeb = Readable.toWeb(process.stdin) as ReadableStream<Uint8Array>
    return ndJsonStream(stdoutWeb, stdinWeb)
  }

  private makeAgent(conn: AgentSideConnection): UpstreamAgent {
    return {
      authenticate: async (_params: AuthenticateRequest): Promise<AuthenticateResponse> => {
        // v0.1 of the SDK does not surface authenticate to user code —
        // agents that need an out-of-band login flow should return an
        // `AUTH_REQUIRED` error from `initialize` instead (§15.6).
        throw new Error('ChannelAgent: authenticate is not supported in v0.1. Surface AUTH_REQUIRED from initialize().')
      },
      cancel: async (notification: CancelNotification): Promise<void> => {
        const state = this.sessions.get(notification.sessionId)
        if (state !== undefined) state.abortController.abort()
        if (this.cancelHandler !== undefined) {
          await this.cancelHandler(notification)
        }
      },
      initialize: async (_params: InitializeRequest): Promise<InitializeResponse> => ({
        agentCapabilities: {
          promptCapabilities: this.config.promptCapabilities,
        },
        agentInfo: {name: this.config.name, version: this.config.version},
        protocolVersion: 1,
      }),
      newSession: async (_params: NewSessionRequest): Promise<NewSessionResponse> => {
        const sessionId = randomUUID()
        this.sessions.set(sessionId, {abortController: new AbortController()})
        return {sessionId}
      },
      prompt: async (params: PromptRequest): Promise<PromptResponse> => {
        if (this.promptHandler === undefined) {
          throw new Error('ChannelAgent: no prompt handler registered. Call agent.onPrompt(...) before agent.run().')
        }

        const state = this.sessions.get(params.sessionId) ?? {abortController: new AbortController()}
        if (!this.sessions.has(params.sessionId)) this.sessions.set(params.sessionId, state)
        const ctx = new PromptContext({
          connection: conn,
          sessionId: params.sessionId,
          signal: state.abortController.signal,
        })
        try {
          return await this.promptHandler(params, ctx)
        } finally {
          ctx._deactivate()
          // Reset the abort controller for the next prompt on this session.
          this.sessions.set(params.sessionId, {abortController: new AbortController()})
        }
      },
    }
  }
}

File: packages/brv-agent-py/src/brv_agent/channel_agent.py
"""ChannelAgent — ergonomic wrapper around the upstream ``acp`` Python lib's
agent protocol. See Channel Protocol §15 for the wire spec this targets.
"""

from __future__ import annotations

import asyncio
import uuid
from typing import Any, Awaitable, Callable, Optional

import acp
from acp.schema import (
    AgentCapabilities,
    ClientCapabilities,
    Implementation,
    InitializeResponse,
    NewSessionResponse,
    PromptCapabilities,
    PromptResponse,
)

from .prompt_context import PromptContext

PromptHandler = Callable[["acp.PromptRequest", PromptContext], Awaitable[PromptResponse]]
CancelHandler = Callable[["acp.CancelNotification"], Awaitable[None] | None]


class ChannelAgent:
    """Build a channel-aware ACP agent in ~30 LOC.

    Example:
        >>> agent = ChannelAgent(name="echo", version="0.1.0",
        ...                       prompt_capabilities={"embedded_context": True})
        >>>
        >>> @agent.on_prompt
        ... async def handle(req, ctx):
        ...     await ctx.send_message_chunk("hi")
        ...     return {"stop_reason": "end_turn"}
        >>>
        >>> import asyncio; asyncio.run(agent.run())
    """

    def __init__(
        self,
        *,
        name: str,
        version: str,
        prompt_capabilities: dict[str, bool] | PromptCapabilities | None = None,
    ) -> None:
        self._name = name
        self._version = version
        if prompt_capabilities is None:
            self._prompt_capabilities = PromptCapabilities()
        elif isinstance(prompt_capabilities, PromptCapabilities):
            self._prompt_capabilities = prompt_capabilities
        else:
            self._prompt_capabilities = PromptCapabilities(**prompt_capabilities)

        self._prompt_handler: PromptHandler | None = None
        self._cancel_handler: CancelHandler | None = None
        # session_id -> asyncio.Event (set when host cancels)
        self._cancel_events: dict[str, asyncio.Event] = {}

    def on_prompt(self, handler: PromptHandler) -> PromptHandler:
        """Decorator: register the prompt handler.

        Usage:
            @agent.on_prompt
            async def handle(req, ctx): ...
        """
        self._prompt_handler = handler
        return handler

    def on_cancel(self, handler: CancelHandler) -> CancelHandler:
        """Decorator: register an optional cancel observer.

        The SDK already triggers ``ctx.signal`` for in-flight prompt
        handlers; ``on_cancel`` is for agents that want to do extra
        bookkeeping (e.g. tear down an external resource) on cancel.
        """
        self._cancel_handler = handler
        return handler

    async def run(
        self,
        *,
        input_stream: Any = None,
        output_stream: Any = None,
    ) -> None:
        """Run the agent loop. Defaults to stdio (NDJSON) — the runner
        used for ``brv channel onboard``. Tests pass explicit
        ``input_stream`` / ``output_stream``.
        """
        await acp.run_agent(
            self,
            input_stream=input_stream,
            output_stream=output_stream,
        )

    # ─── acp.Agent protocol implementation ──────────────────────────────

    async def initialize(
        self,
        protocol_version: int,
        client_capabilities: ClientCapabilities | None = None,
        client_info: Optional[Implementation] = None,
        **kwargs: Any,
    ) -> InitializeResponse:
        return InitializeResponse(
            protocolVersion=1,
            agentCapabilities=AgentCapabilities(
                promptCapabilities=self._prompt_capabilities,
            ),
            agentInfo=Implementation(name=self._name, version=self._version),
        )

    async def new_session(
        self,
        cwd: str,
        mcp_servers: Any = None,
        **kwargs: Any,
    ) -> NewSessionResponse:
        session_id = str(uuid.uuid4())
        self._cancel_events[session_id] = asyncio.Event()
        return NewSessionResponse(sessionId=session_id)

    async def prompt(
        self,
        prompt: list[Any],
        session_id: str,
        **kwargs: Any,
    ) -> PromptResponse:
        if self._prompt_handler is None:
            raise RuntimeError(
                "ChannelAgent: no prompt handler registered. Use @agent.on_prompt before agent.run()."
            )

        cancel_event = self._cancel_events.setdefault(session_id, asyncio.Event())
        ctx = PromptContext(
            client=_get_client(self),
            session_id=session_id,
            cancel_event=cancel_event,
        )
        try:
            result = await self._prompt_handler(
                _build_prompt_request(prompt=prompt, session_id=session_id),
                ctx,
            )
        finally:
            ctx._deactivate()
            # Review fix #10: drop the cancel event once the handler has
            # resolved. The Event only matters DURING a prompt; keeping a
            # stale entry forever (or recreating one for a "next prompt"
            # that may never come) is the leak. A subsequent `prompt` on
            # the same `session_id` recreates the entry just-in-time via
            # the `setdefault(...)` above.
            self._cancel_events.pop(session_id, None)

        if isinstance(result, PromptResponse):
            return result
        if isinstance(result, dict):
            stop_reason = result.get("stop_reason") or result.get("stopReason") or "end_turn"
            # StopReason in the upstream schema is a `Literal` alias — pass
            # the bare string and let pydantic validate.
            return PromptResponse(stopReason=stop_reason)  # type: ignore[arg-type]
        raise TypeError(
            "ChannelAgent: on_prompt handler must return PromptResponse or "
            "{'stop_reason': '...'}"
        )

    async def cancel(self, session_id: str, **kwargs: Any) -> None:
        event = self._cancel_events.get(session_id)
        if event is not None:
            event.set()
        if self._cancel_handler is not None:
            from acp import CancelNotification
            notification = CancelNotification(sessionId=session_id)
            maybe_awaitable = self._cancel_handler(notification)
            if asyncio.iscoroutine(maybe_awaitable):
                await maybe_awaitable

    # The upstream acp Protocol requires `authenticate` to satisfy the
    # interface. v0.1 of brv-agent doesn't surface auth to user code — agents
    # that need an out-of-band login flow should raise an `AUTH_REQUIRED`
    # error from `initialize()` instead (§15.6).
    async def authenticate(self, method_id: str, **kwargs: Any) -> None:
        raise RuntimeError(
            "ChannelAgent: authenticate is not supported in v0.1. "
            "Surface AUTH_REQUIRED from initialize() instead."
        )

    # `on_connect` is called by acp.run_agent with the Client object so we
    # can call back into it (session_update, request_permission). We stash
    # it for PromptContext to use.
    def on_connect(self, client: acp.Client) -> None:
        self._client = client


# Internal helpers (separated for testability).

def _get_client(agent: "ChannelAgent") -> acp.Client:
    client = getattr(agent, "_client", None)
    if client is None:
        raise RuntimeError(
            "ChannelAgent.prompt() called before on_connect — did you call agent.run()?"
        )
    return client


def _build_prompt_request(*, prompt: list[Any], session_id: str) -> Any:
    """Wrap the raw prompt block list into a small adapter object the
    handler can pattern-match on. Mirrors the TS SDK's
    `PromptRequest` shape so the docs read identically across languages.
    """
    class _Req:
        def __init__(self) -> None:
            self.prompt = prompt
            self.session_id = session_id

    return _Req()
